Add MIT Press translator

GuyAglionby opened the pull request:
As title. Only thing I'm not sure about is how to handle the book subtitle. For example, for title "Elements of Causal Inference" with subtitle "Foundations and Learning Algorithms", would the appropriate `title` be "Elements of Causal Inference: Foundations and Learning Algorithms" and `shortTitle` "Elements of Causal Inference"? Or should the `title` just be "Elements of Causal Inference"

Thanks!

adam3smith commented on MIT Press.js:
```diff
@@ -0,0 +1,206 @@
+{
+	"translatorID": "ac277fbe-000c-46da-b145-fbe799d17eda",
+	"label": "MIT Press",
```
Should be called "MIT Press Books" to distinguish it from existing "MIT Press Journals" translator

adam3smith commented on MIT Press.js:
```diff
@@ -0,0 +1,206 @@
+	"translatorID": "ac277fbe-000c-46da-b145-fbe799d17eda",
+	"label": "MIT Press",
+	"creator": "Guy Aglionby",
+	"target": "https://(www\\.)?mitpress\\.mit\\.edu/",
```
again, to distinguish it from journals, make the regex more restrictive (just re-using the bits from detect Web, which you may then be able to simplify)

adam3smith commented on MIT Press.js:
```diff
@@ -0,0 +1,206 @@
+
+function scrape(doc, url) {
+	let authorsElement = ZU.xpath(doc, '//span[@class = "book__authors"]/p[1]/a');
+	let authors = authorsElement.map(function (author) {
```
Could we try to distinguish editors by using the "Edited by" wording, e.g. https://mitpress.mit.edu/books/construction-site-possible-worlds

adam3smith replied:
Example of book with editor and author: https://mitpress.mit.edu/books/kings-and-things
Example of book with author and translator: https://mitpress.mit.edu/books/ribbon-olympias-throat

adam3smith commented on MIT Press.js:
```diff
@@ -0,0 +1,206 @@
+	item.place = 'Cambridge, MA, USA';
+	item.publisher = 'MIT Press';
+	item.language = 'en';
+	item.title = ZU.xpathText(doc, '//h1[@class = "book__title"]').trim();
```
I think we do want the subtitle after : as per your question. E.g. for the causal inference book, that's how it is cataloged by LoC.

GuyAglionby commented:
Thanks for the feedback! Made changes as requested. Added some new tests to include the multiple contributor types + cases where two elements for the book series were present.

zuphilip commented on MIT Press Books.js:
```diff
@@ -0,0 +1,298 @@
+	let title = ZU.xpathText(doc, '//h1[@class = "book__title"]').trim();
+	let subtitle = ZU.xpathText(doc, '//h2[@class = "book__subtitle"]').trim();
+	if (subtitle) {
+		item.title = [title, subtitle].join(': ');
```
Move the `trim()` from the `subtitle` inside the `if` block, because otherwise there is a potential that you try to trim from `null` which will fail.

zuphilip commented on MIT Press Books.js:
```diff
@@ -0,0 +1,298 @@
+				"tags": [],
+				"notes": [],
+				"seeAlso": []
+			}
```
You could also add the short abstract from the `div` of class `book__blurb`.

zuphilip commented on MIT Press Books.js:
```diff
@@ -0,0 +1,298 @@
+				"numPages": "136",
+				"place": "Cambridge, MA, USA",
+				"publisher": "MIT Press",
+				"series": "Urbanomic",
```
Actually, this looks like we should better save this as the publisher info, cf. https://www.urbanomic.com/book/construction-site-possible-worlds/ .

zuphilip commented on MIT Press Books.js:
```diff
@@ -0,0 +1,298 @@
+		"items": [
+			{
+				"itemType": "book",
+				"title": "The Sciences of the Artificial, Reissue Of The Third Edition With A New Introduction By John Laird",
```
It looks like the edition (as well as volume) info is saved as part of the title. This would be good to try to split it of, if we can. This example here looks extremely challenging and therefore I am not sure that we should tackle it at all. But easier example are maybe
- Foundations of Machine Learning, Second Edition: https://mitpress.mit.edu/books/foundations-machine-learning-second-edition
- Collapse, Volume 8: https://mitpress.mit.edu/books/collapse-volume-8
- Introduction to Algorithms, Third Edition: https://mitpress.mit.edu/books/introduction-algorithms-third-edition

Can we fix at least these ans similar examples?

GuyAglionby commented:
Thanks for the comments and sorry for the delay here. I've updated this so that the publisher and series are appropriately set, and included the blurb as an abstract. 

I had a look on the site and they're very regular on how volumes and editions are noted -- they're always of the form `Book Title, Volume \d+` or `Book Title, [\w ]+Edition([\w ]+)?`. Volume I just extracted the number; for edition I had a look at what the other translators do and only include "Edition" if there's text after it (e.g. [Reissue Of The Third Edition With A New Introduction By John Laird](https://github.com/zotero/translators/pull/2043/commits/8a7195b94df2db6fc339e3589558ed1fc6a13fcb#diff-58d461527180df476ba1d3d619f51ea5R257)), otherwise exclude it (e.g. [Revised And Expanded](https://github.com/zotero/translators/pull/2043/commits/50d0ea1936778cbe96048a8e68d971d458b4532d#diff-58d461527180df476ba1d3d619f51ea5R516)). This is easily changed if another way would be preferable. In all cases all the text is removed from the title.

GuyAglionby commented:
@zuphilip I hope the changes address your concerns -- please let me know if there's anything else that needs doing!

GuyAglionby commented:
@adam3smith @zuphilip please let me know if there's anything else!

boan-anbo commented:
Bumping this as I was checking if anyone already made the MIT press translator. It's used pretty frequently. If there is anything I can help on top of @GuyAglionby's PR let me know.

AbeJellinek commented:
Sorry for the delay on this. Great translator, and very much needed. I tweaked the edition handling so that we get numbers, not spelled-out ordinals, and added series editors. I also removed a 404'd test.

AbeJellinek commented:
I think all is good @boan-anbo, thanks for offering help! I've just been slow to get to this one.
